libr/bin/format/xtac/bin_xtac.c

```diff
@@ -0,0 +1,587 @@
+	rbin->cb_printf ("  ptr_to_mod_name : 0x%x\n", bin->header->ptr_to_mod_name);
+	rbin->cb_printf ("  size_of_mod_name : 0x%x\n", bin->header->size_of_mod_name);
+	rbin->cb_printf ("  module name : %s\n", bin->mod_name_u8);
+
+	rbin->cb_printf ("  ptr_to_nt_pname : 0x%x\n", bin->header->ptr_to_nt_pname);
+	rbin->cb_printf ("  size_of_nt_pname : 0x%x\n", bin->header->size_of_nt_pname);
+	rbin->cb_printf ("  module name : %s\n", bin->nt_path_name_u8);
+
+	rbin->cb_printf ("  ptr_to_head_blck_stub : 0x%x\n", bin->header->ptr_to_head_blck_stub);
+	rbin->cb_printf ("  ptr_to_tail_blck_stub : 0x%x\n", bin->header->ptr_to_tail_blck_stub);
+	rbin->cb_printf ("  size_of_blck_stub_code : 0x%x\n", bin->header->size_of_blck_stub_code);
+	rbin->cb_printf ("  ptr_to_xtac_linked_list_head : 0x%x\n", bin->header->ptr_to_xtac_linked_list_head);
+	rbin->cb_printf ("  ptr_to_xtac_linked_list_tail : 0x%x\n", bin->header->ptr_to_xtac_linked_list_tail);
+
+	rbin->cb_printf ("address pairs (x86, arm64):\n");
+	const ut32 n_addr_pairs = bin->header->num_of_addr_pairs;
+	for (ut32 i = 0; i < n_addr_pairs; i++) {
+		rbin->cb_printf ("  0x%x, 0x%x\n", bin->address_pairs[i].x86_rva, bin->address_pairs[i].arm64_rva);
+	}
+
+	rbin->cb_printf ("blck stubs:\n");
+	RBinBlckStubHeader *blck_stub = NULL;
+	RListIter *iter = NULL;
+	r_list_foreach (bin->blck_stubs, iter, blck_stub) {
+		rbin->cb_printf ("  blck stub entry\n");
+		rbin->cb_printf ("    ptr_to_entry : 0x%x\n", blck_stub->ptr_to_entry);
+		rbin->cb_printf ("    magic : 0x%x\n", blck_stub->magic);
+		rbin->cb_printf ("    offset_to_next_entry : 0x%x\n", blck_stub->offset_to_next_entry);
+		rbin->cb_printf ("    ptr_to_next_entry : 0x%x\n", blck_stub->ptr_to_next_entry);
+		rbin->cb_printf ("    padding : 0x%x\n", blck_stub->padding);
+	}
+
+	rbin->cb_printf ("xtac linked list:\n");
+	RBinXtacLinkedListEntry *entry = NULL;
+	r_list_foreach (bin->xtac_linked_list, iter, entry) {
+		rbin->cb_printf ("  xtac linked list entry\n");
+		rbin->cb_printf ("    ptr_to_entry : 0x%x\n", entry->ptr_to_entry);
+		rbin->cb_printf ("    meta_data : 0x%x\n", GET_META_DATA (entry->meta_and_offset));
+		rbin->cb_printf ("    offset_to_next_entry : 0x%x\n", GET_OFFSET (entry->meta_and_offset) * 4);
+		rbin->cb_printf ("    forward_edge_addr : 0x%x\n", entry->forward_edge_addr);
+		rbin->cb_printf ("    backward_edge_addr : 0x%x\n", entry->backward_edge_addr);
+	}
+}
+
+static bool r_bin_xtac_read_header(RBinXtacObj *bin) {
+	if (!(bin->header = R_NEW0 (RBinXtacHeader))) {
+		r_sys_perror ("R_NEW0 (xtac header)");
+		return false;
+	}
+	if (r_buf_read_at (bin->b, 0, (ut8*) bin->header, sizeof (RBinXtacHeader)) < 0) {
+		R_LOG_WARN ("Warning: read (xtac header)\n");
+		free (bin->header);
+		bin->header = NULL;
+		return false;
+	}
+	return true;
+}
+
+static bool r_bin_xtac_read_address_pairs(RBinXtacObj *bin) {
+	const ut32 n_addr_pairs = bin->header->num_of_addr_pairs;
+	const ut32 addr_pair_size = n_addr_pairs * sizeof (X86ArmAddrPair);
```

Comment:
this can overflow ut32 on fuzzed files use the R_OVF macros to detect the overflow or add a reasonable limit.

Reply:
What are R_OVF macros? I'll review the radare2 code base, but I cannot find these.

Reply:
https://github.com/radareorg/radare2/blob/master/libr/include/r_types_overflow.h